In [1]:
import pandas as pd, numpy as np, matplotlib.pyplot as plt, seaborn as sns

In [2]:
# Load all 4 datasets 
company_info = pd.read_csv("../data/raw/company_info.csv")
market_indices = pd.read_csv("../data/raw/market_indices.csv")
stock_prices = pd.read_csv("../data/raw/stock_prices.csv")
stock_prices_indicators = pd.read_csv("../data/raw/stock_prices_with_indicators.csv")
print (" All CSVc loaded successfully (if no errors shown).")

 All CSVc loaded successfully (if no errors shown).


In [4]:
datasets = {
    "company_info": company_info,
    "market_indices": market_indices,
    "stock_prices": stock_prices,
    "stock_prices_indicators": stock_prices_indicators
}

### 1. Company_info dataset 

In [5]:
company_info.shape

(20, 4)

In [6]:
company_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   ticker        20 non-null     object
 1   company_name  20 non-null     object
 2   sector        20 non-null     object
 3   ipo_date      20 non-null     object
dtypes: object(4)
memory usage: 772.0+ bytes


In [7]:
company_info.head()

,ticker,company_name,sector,ipo_date
0,STK001,TechCorp,Technology,2021-04-10
1,STK002,DataSystems,Technology,2016-10-12
2,STK003,CloudNine,Technology,2016-01-18
3,STK004,CyberShield,Technology,2022-02-21
4,STK005,MediPharm,Healthcare,2018-03-05


#### Insight 
Company_Info dataset has 20 rows and 4 columns with no null values. This file help to map tickers to company names and sectors for grouping, filtering, and sector-based analysis.

### 2. Market_Indices dataset

In [8]:
market_indices.shape

(780, 7)

In [9]:
market_indices.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 780 entries, 0 to 779
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   date           780 non-null    object 
 1   sp500_close    780 non-null    float64
 2   nasdaq_close   780 non-null    float64
 3   vix_close      780 non-null    float64
 4   treasury_10y   780 non-null    float64
 5   dollar_index   780 non-null    float64
 6   market_regime  780 non-null    object 
dtypes: float64(5), object(2)
memory usage: 42.8+ KB


In [10]:
market_indices.head()

,date,sp500_close,nasdaq_close,vix_close,treasury_10y,dollar_index,market_regime
0,2021-01-04,3891.04,12952.02,20.48,1.496,90.07,bull
1,2021-01-05,3847.20,13116.52,19.89,1.504,90.73,bull
2,2021-01-06,3878.24,12818.48,20.02,1.507,90.46,bull
3,2021-01-07,3803.95,13290.11,19.59,1.498,90.67,bull
4,2021-01-08,3715.64,13365.73,20.14,1.498,90.64,bull


#### Insight 
Market Indices has 780 rows and 7 columns with no null entires 

### 3. Stock_Prices dataset 

In [11]:
stock_prices.shape

(15600, 8)

In [12]:
stock_prices.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15600 entries, 0 to 15599
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   ticker          15600 non-null  object 
 1   date            15600 non-null  object 
 2   open            15600 non-null  float64
 3   high            15600 non-null  float64
 4   low             15600 non-null  float64
 5   close           15600 non-null  float64
 6   volume          15600 non-null  int64  
 7   adjusted_close  15600 non-null  float64
dtypes: float64(5), int64(1), object(2)
memory usage: 975.1+ KB


In [13]:
stock_prices.head()

,ticker,date,open,high,low,close,volume,adjusted_close
0,STK001,2021-01-04,158.09,160.97,158.09,160.11,962644,160.11
1,STK001,2021-01-05,163.16,165.50,160.76,162.36,1312685,162.36
2,STK001,2021-01-06,161.89,162.51,160.94,161.78,1449177,161.78
3,STK001,2021-01-07,163.33,167.90,163.33,167.07,1534833,167.07
4,STK001,2021-01-08,168.20,168.20,164.12,165.68,848261,165.68


#### Insight 
Stock_prices dataset has 15600 rows and 8 columns with no null entires 

### 4. Stock_Prices_Indicators

In [14]:
stock_prices_indicators.shape

(15502, 31)

In [15]:
stock_prices_indicators.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15502 entries, 0 to 15501
Data columns (total 31 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   ticker            15502 non-null  object 
 1   date              15502 non-null  object 
 2   open              15502 non-null  float64
 3   high              15502 non-null  float64
 4   low               15502 non-null  float64
 5   close             15502 non-null  float64
 6   volume            15502 non-null  float64
 7   adjusted_close    15502 non-null  float64
 8   sma_20            15192 non-null  float64
 9   sma_50            15502 non-null  float64
 10  sma_200           15502 non-null  float64
 11  ema_12            15502 non-null  float64
 12  ema_26            15502 non-null  float64
 13  macd              15191 non-null  float64
 14  macd_signal       15502 non-null  float64
 15  macd_histogram    15502 non-null  float64
 16  rsi_14            15172 non-null  float6

In [16]:
stock_prices_indicators.head()

,ticker,date,open,high,low,close,volume,adjusted_close,sma_20,sma_50,...,true_range,atr_14,volume_sma_20,volume_ratio,momentum_10,momentum_20,price_to_sma_50,volatility_20,future_return_5d,trend_label
0,STK001,2021-01-04,158.09,160.97,158.09,160.11,962644.0,160.11,160.110000,160.110000,...,2.88,2.880,962644.00,1.000000,NaN,NaN,0.000000,NaN,0.026357,Uptrend
1,STK001,2021-01-05,163.16,165.50,160.76,162.36,1312685.0,162.36,161.235000,161.235000,...,5.39,4.135,1137664.50,1.153842,NaN,NaN,0.006977,NaN,0.018169,Sideways
2,STK001,2021-01-06,161.89,162.51,160.94,161.78,1449177.0,161.78,161.416667,161.416667,...,1.57,3.280,1241502.00,1.167277,NaN,NaN,0.002251,0.012463,-0.012301,Sideways
3,STK001,2021-01-07,163.33,167.90,163.33,167.07,1534833.0,167.07,162.830000,162.830000,...,6.12,3.990,1314834.75,1.167320,NaN,NaN,0.026039,0.018138,-0.061950,Downtrend
4,STK001,2021-01-08,168.20,168.20,164.12,165.68,848261.0,165.68,163.400000,163.400000,...,4.08,4.008,1221520.00,0.694431,NaN,NaN,0.013953,0.018663,-0.070196,Downtrend


#### Insight 
stock_prices_indicators has 15502 rows, 31 columns and some null values

In [17]:
total_missing = stock_prices_indicators.isnull().sum().sum()
total_cells = stock_prices_indicators.shape[0] * stock_prices_indicators.shape[1]
overall_missing_percent = (total_missing / total_cells) * 100

print(f"Overall missing percentage: {overall_missing_percent:.2f}%")


Overall missing percentage: 0.47%


In [18]:
stock_prices_indicators.head()

,ticker,date,open,high,low,close,volume,adjusted_close,sma_20,sma_50,...,true_range,atr_14,volume_sma_20,volume_ratio,momentum_10,momentum_20,price_to_sma_50,volatility_20,future_return_5d,trend_label
0,STK001,2021-01-04,158.09,160.97,158.09,160.11,962644.0,160.11,160.110000,160.110000,...,2.88,2.880,962644.00,1.000000,NaN,NaN,0.000000,NaN,0.026357,Uptrend
1,STK001,2021-01-05,163.16,165.50,160.76,162.36,1312685.0,162.36,161.235000,161.235000,...,5.39,4.135,1137664.50,1.153842,NaN,NaN,0.006977,NaN,0.018169,Sideways
2,STK001,2021-01-06,161.89,162.51,160.94,161.78,1449177.0,161.78,161.416667,161.416667,...,1.57,3.280,1241502.00,1.167277,NaN,NaN,0.002251,0.012463,-0.012301,Sideways
3,STK001,2021-01-07,163.33,167.90,163.33,167.07,1534833.0,167.07,162.830000,162.830000,...,6.12,3.990,1314834.75,1.167320,NaN,NaN,0.026039,0.018138,-0.061950,Downtrend
4,STK001,2021-01-08,168.20,168.20,164.12,165.68,848261.0,165.68,163.400000,163.400000,...,4.08,4.008,1221520.00,0.694431,NaN,NaN,0.013953,0.018663,-0.070196,Downtrend


In [19]:
stock_prices_indicators.sort_values("date", inplace=True)
stock_prices_indicators.ffill(inplace=True)
stock_prices_indicators.bfill(inplace=True)


In [20]:
stock_prices_indicators.head()

,ticker,date,open,high,low,close,volume,adjusted_close,sma_20,sma_50,...,true_range,atr_14,volume_sma_20,volume_ratio,momentum_10,momentum_20,price_to_sma_50,volatility_20,future_return_5d,trend_label
0,STK001,2021-01-04,158.09,160.97,158.09,160.11,962644.0,160.11,160.11,160.11,...,2.88,2.88,962644.0,1.0,-0.044552,0.156698,0.0,0.020763,0.026357,Uptrend
2325,STK004,2021-01-04,151.44,155.22,150.33,152.00,963623.0,152.00,152.00,152.00,...,4.89,4.89,963623.0,1.0,-0.044552,0.156698,0.0,0.020763,-0.032829,Downtrend
3100,STK005,2021-01-04,89.08,90.46,88.01,89.70,1491126.0,89.70,89.70,89.70,...,2.45,2.45,1491126.0,1.0,-0.044552,0.156698,0.0,0.020763,-0.011594,Sideways
3875,STK006,2021-01-04,84.61,85.19,82.84,84.13,954683.0,84.13,84.13,84.13,...,2.35,2.35,954683.0,1.0,-0.044552,0.156698,0.0,0.020763,0.032093,Uptrend
4650,STK007,2021-01-04,76.13,77.80,76.00,76.85,947006.0,76.85,76.85,76.85,...,1.80,1.80,947006.0,1.0,-0.044552,0.156698,0.0,0.020763,0.008068,Sideways


In [21]:
stock_prices_indicators.isnull().sum()

ticker              0
date                0
open                0
high                0
low                 0
close               0
volume              0
adjusted_close      0
sma_20              0
sma_50              0
sma_200             0
ema_12              0
ema_26              0
macd                0
macd_signal         0
macd_histogram      0
rsi_14              0
bb_middle           0
bb_upper            0
bb_lower            0
bb_width            0
true_range          0
atr_14              0
volume_sma_20       0
volume_ratio        0
momentum_10         0
momentum_20         0
price_to_sma_50     0
volatility_20       0
future_return_5d    0
trend_label         0
dtype: int64

In [22]:
# Identify datatype mismatches 
for name, df in datasets.items():
    print(f"\n📘 Data Types in {name}")
    display(df.dtypes)



📘 Data Types in company_info


ticker          object
company_name    object
sector          object
ipo_date        object
dtype: object


📘 Data Types in market_indices


date              object
sp500_close      float64
nasdaq_close     float64
vix_close        float64
treasury_10y     float64
dollar_index     float64
market_regime     object
dtype: object


📘 Data Types in stock_prices


ticker             object
date               object
open              float64
high              float64
low               float64
close             float64
volume              int64
adjusted_close    float64
dtype: object


📘 Data Types in stock_prices_indicators


ticker               object
date                 object
open                float64
high                float64
low                 float64
close               float64
volume              float64
adjusted_close      float64
sma_20              float64
sma_50              float64
sma_200             float64
ema_12              float64
ema_26              float64
macd                float64
macd_signal         float64
macd_histogram      float64
rsi_14              float64
bb_middle           float64
bb_upper            float64
bb_lower            float64
bb_width            float64
true_range          float64
atr_14              float64
volume_sma_20       float64
volume_ratio        float64
momentum_10         float64
momentum_20         float64
price_to_sma_50     float64
volatility_20       float64
future_return_5d    float64
trend_label          object
dtype: object

In [23]:
# Fixing object type for date columns 
for name, df in datasets.items():
    # Convert 'date' column if it exists
    if 'date' in df.columns:
        try:
            df['date'] = pd.to_datetime(df['date'], errors='coerce')
            print(f" Converted 'date' to datetime in {name}")
        except Exception as e:
            print(f" Error converting 'date' in {name}: {e}")
    
    # Convert 'ipo_date' column if it exists
    if 'ipo_date' in df.columns:
        try:
            df['ipo_date'] = pd.to_datetime(df['ipo_date'], errors='coerce')
            print(f" Converted 'ipo_date' to datetime in {name}")
        except Exception as e:
            print(f" Error converting 'ipo_date' in {name}: {e}")


 Converted 'ipo_date' to datetime in company_info
 Converted 'date' to datetime in market_indices
 Converted 'date' to datetime in stock_prices
 Converted 'date' to datetime in stock_prices_indicators


In [24]:
# Check for duplicate 
for name, df in datasets.items():
    duplicates = df.duplicated().sum()
    print(f"{name}: {duplicates} duplicate rows")


company_info: 0 duplicate rows
market_indices: 0 duplicate rows
stock_prices: 0 duplicate rows
stock_prices_indicators: 0 duplicate rows


In [26]:
missing_dates_report = []

for ticker in stock_prices_indicators['ticker'].unique():
    sample = stock_prices_indicators[stock_prices_indicators['ticker'] == ticker].sort_values('date')
    
    # Generate expected business-day range
    expected_dates = pd.date_range(sample['date'].min(), sample['date'].max(), freq='B')
    
    # Compare with actual dates
    missing_dates = expected_dates.difference(sample['date'])
    
    # Store result
    missing_dates_report.append({'ticker': ticker, 'missing_dates': len(missing_dates)})

# Convert to DataFrame for easy viewing
missing_dates_df = pd.DataFrame(missing_dates_report)
print(missing_dates_df)


    ticker  missing_dates
0   STK001              0
1   STK004              0
2   STK005              0
3   STK006              0
4   STK007              0
5   STK008              0
6   STK009              0
7   STK010              0
8   STK003              0
9   STK012              0
10  STK014              0
11  STK015              0
12  STK016              0
13  STK017              0
14  STK018              0
15  STK019              0
16  STK020              0
17  STK013              0
18  STK002              0
19  STK011              0


In [27]:
#  Verify date ranges and continuity in time-series data
for name, df in [("market_indices", market_indices),  ("stock_prices", stock_prices), ("stock_prices_indicators", stock_prices_indicators)]:
    if 'date' in df.columns:
        print(f"\n📅 {name} Date Range:")
        print("Start:", df['date'].min())
        print("End:", df['date'].max())
        print("Unique Dates:", df['date'].nunique())



📅 market_indices Date Range:
Start: 2021-01-04 00:00:00
End: 2023-12-29 00:00:00
Unique Dates: 780

📅 stock_prices Date Range:
Start: 2021-01-04 00:00:00
End: 2023-12-29 00:00:00
Unique Dates: 780

📅 stock_prices_indicators Date Range:
Start: 2021-01-04 00:00:00
End: 2023-12-25 00:00:00
Unique Dates: 776


### Insight 
The first two datasets (market_indices and stock_prices) have the same:
 1. Start and end dates
 2. Total number of trading days (780)
That means they cover the same full date range — perfect.

But stock_prices_indicators ends on 2023-12-25, not 2023-12-29, and has 776 unique dates, i.e. 4 fewer days.

What This Means?

The indicator dataset is slightly shorter — it didn’t generate indicators (like moving averages or RSI) for the final few days.
This is common because technical indicators (e.g., SMA, RSI, MACD) rely on rolling windows — the last few rows sometimes have missing values that get dropped during computation or cleaning.

In [28]:
# let align to the same date range 
full_range = pd.date_range('2021-01-04', '2023-12-29', freq='B')

aligned_data = []

for ticker, group in stock_prices_indicators.groupby('ticker'):
    group = (
        group.set_index('date')
        .reindex(full_range)
        .ffill()
        .reset_index()
        .rename(columns={'index': 'date'})
    )
    group['ticker'] = ticker
    aligned_data.append(group)

stock_prices_indicators = pd.concat(aligned_data, ignore_index=True)



In [29]:
#  confirm the date ranges alignment and continuity in time-series data
for name, df in [("market_indices", market_indices),  ("stock_prices", stock_prices), ("stock_prices_indicators", stock_prices_indicators)]:
    if 'date' in df.columns:
        print(f"\n📅 {name} Date Range:")
        print("Start:", df['date'].min())
        print("End:", df['date'].max())
        print("Unique Dates:", df['date'].nunique())


📅 market_indices Date Range:
Start: 2021-01-04 00:00:00
End: 2023-12-29 00:00:00
Unique Dates: 780

📅 stock_prices Date Range:
Start: 2021-01-04 00:00:00
End: 2023-12-29 00:00:00
Unique Dates: 780

📅 stock_prices_indicators Date Range:
Start: 2021-01-04 00:00:00
End: 2023-12-29 00:00:00
Unique Dates: 780


🧾 **Data Quality Summary and Insights**

| Dataset | Missing Values (%) | Key Fixes/Actions | Data Type Issues | Date Range | Insights |
|----------|--------------------|------------------|------------------|-------------|-----------|
| **company_info** | 0.00% | Converted `ipo_date` to datetime | Initially stored as object | N/A | Clean metadata for companies; ready for merging or reference use. |
| **market_indices** | 0.00% | Converted `date` to datetime | Initially stored as object | 2021-01-04 → 2023-12-29 (780 unique dates) | Fully continuous business-day data; no missing values. Suitable as external market context. |
| **stock_prices** | 0.00% | Converted `date` to datetime | Initially stored as object | 2021-01-04 → 2023-12-29 (780 unique dates) | Complete daily price data across tickers; aligned with market_indices for temporal analysis. |
| **stock_prices_indicators** | ~0.47% overall (before fix) | Converted `date` to datetime; reindexed each ticker to full range; forward-filled gaps | Initially stored as object | 2021-01-04 → 2023-12-29 (780 unique dates after alignment) | Technical indicators had minor gaps (mostly due to rolling windows). Forward-filled values resolved missing points; dataset now consistent for modeling. |


## Insight: 
Data Type Cleaning:
All date and ipo_date columns were standardized to datetime format, resolving prior type inconsistencies.

Temporal Consistency:
All datasets now share the same trading-day range (780 business days), enabling synchronized time-series modeling.

Minimal Missing Data:
The few missing values in technical indicators were expected (rolling-window calculations).
After forward-filling, continuity is restored with negligible data loss risk.

Ready for Analysis:
Datasets are well-structured for feature engineering, correlation studies, and predictive modeling (e.g., stock trend classification).

In [30]:
modeling_df = stock_prices_indicators.copy()

In [31]:
modeling_df.head()

,date,ticker,open,high,low,close,volume,adjusted_close,sma_20,sma_50,...,true_range,atr_14,volume_sma_20,volume_ratio,momentum_10,momentum_20,price_to_sma_50,volatility_20,future_return_5d,trend_label
0,2021-01-04,STK001,158.09,160.97,158.09,160.11,962644.0,160.11,160.110000,160.110000,...,2.88,2.880,962644.00,1.000000,-0.044552,0.156698,0.000000,0.020763,0.026357,Uptrend
1,2021-01-05,STK001,163.16,165.50,160.76,162.36,1312685.0,162.36,161.235000,161.235000,...,5.39,4.135,1137664.50,1.153842,-0.044552,0.156698,0.006977,0.020763,0.018169,Sideways
2,2021-01-06,STK001,161.89,162.51,160.94,161.78,1449177.0,161.78,161.416667,161.416667,...,1.57,3.280,1241502.00,1.167277,-0.044552,0.156698,0.002251,0.012463,-0.012301,Sideways
3,2021-01-07,STK001,163.33,167.90,163.33,167.07,1534833.0,167.07,162.830000,162.830000,...,6.12,3.990,1314834.75,1.167320,-0.044552,0.156698,0.026039,0.018138,-0.061950,Downtrend
4,2021-01-08,STK001,168.20,168.20,164.12,165.68,848261.0,165.68,163.400000,163.400000,...,4.08,4.008,1221520.00,0.694431,-0.044552,0.156698,0.013953,0.018663,-0.070196,Downtrend


In [32]:
modeling_df.isnull().sum()

date                0
ticker              0
open                0
high                0
low                 0
close               0
volume              0
adjusted_close      0
sma_20              0
sma_50              0
sma_200             0
ema_12              0
ema_26              0
macd                0
macd_signal         0
macd_histogram      0
rsi_14              0
bb_middle           0
bb_upper            0
bb_lower            0
bb_width            0
true_range          0
atr_14              0
volume_sma_20       0
volume_ratio        0
momentum_10         0
momentum_20         0
price_to_sma_50     0
volatility_20       0
future_return_5d    0
trend_label         0
dtype: int64

In [33]:
# let save our clean data for modeling 
modeling_df.to_csv("../data/processed/modeling_df.csv", index=False)